### PR TITLE
Build revisions

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -38,4 +38,4 @@ src/main.o: $(asm_files) $(gfx_files:.png=.2bpp) $(bin_files)
 # Generating a mapfile is required thanks to a bug in rgblink.
 game.gbc: $(obj)
 	rgblink -n $*.sym -m $*.map -o $@ $(obj)
-	rgbfix  -c -n 0 -r 0x03 -s -l 0x33 -k "01" -m 0x1B -j -p 0xFF -t "ZELDA" -v $@
+	rgbfix  -c -n 1 -r 0x03 -s -l 0x33 -k "01" -m 0x1B -j -p 0xFF -t "ZELDA" -v $@
```
Comment (kemenaran) [inline Makefile:41]: For the Makefile, if you want to have it build several different revisions, you way want to have a look at how [pokered's Makefile](https://github.com/pret/pokered/blob/master/Makefile) does to build two different version (red and blue).

Basically they have two targets, `red.gbc` and `blue.gbc`, that build the two variants (and also checks the checksum for the two variants).

I guess here we could have `game_rev0.gbc` and `game_rev1.gbc`. `game_rv1.gbc` would build the game with an external symbol `BUILD_REV` defined (using something like `rgbasm -D BUILD_REV r1`), which we could then use to write in the assembly `IF BUILD_REV == "r1"`.

**Comment by kemenaran** [comment]
That's quite impressive 👍 

So they aren't *that* many changes between rev0 and rev1 after all…

Do you feel like you could adjust the build to be able to compile the two different revisions?

File: src/code/entities/bank36.asm [hunk 2]
```diff
@@ -8267,6 +8286,53 @@ EntityInitGenie::
     ld   [hl], a                                  ; $72B8: $77
     ret                                           ; $72B9: $C9
 
+fn_036_72d5:
+    ldh a, [hActiveEntityState]
+    sub $02
+    rst $00
```
Comment (marijnvdwerf) [inline src/code/entities/bank36.asm:8292]: Oh, this is a jump table

**Comment by Xkeeper0** [comment]
I think my only suggestion would be either using `game.gbc` and `game-revA.gbc`, or (if other language versions are going to be made, too) using longer filenames

`ZLADX EN 1.0.gbc`, `ZLADX JP 1.2.gbc`, that sort of thing. This isn't concrete, though, just a thought.

File: src/options.asm [hunk 3]
```diff
@@ -44,4 +44,166 @@
...
+__PATCH_3__ = FALSE ; Applies to JP1+, EN1+
+__PATCH_4__ = FALSE ; Applies to DE1+, JP2+, EN2+, FR1+
+__PATCH_5__ = FALSE ; Applies to DE
+__PATCH_6__ = FALSE ; Applies to JP
+__PATCH_7__ = FALSE ; Applies to FR
+__PATCH_8__ = FALSE ; Applies to JP1+
+__PATCH_9__ = FALSE ; Applies to DE, JP
+__PATCH_A__ = 0 ; 1 in JP, 2 in DE
+__PATCH_B__ = 0 ; 1 in JP/DE, 2 in FR
+__PATCH_C__ = FALSE ; Applies to EN
```
Comment (kemenaran) [inline src/options.asm:68]: I feel like these `__PATCH_XXX___` a bit confusing. The names don't really convey meaning, so I don't know how the patches relates to revisions.

Ideally, there would be some wa to give them meaning – like `__REVISION_2__`, or `LANG_FR`.

I'd say that:

- Probably the `_PATCH_XXX_` constants that only apply to a single language (like `__PATCH_7__`) can be converted to an `IF` right in the source code (like `IF LANG == FR`).
- Maybe some patches can be converted to a revision level instead.

File: src/options.asm [hunk 4]
```diff
@@ -44,4 +44,166 @@
...
+FILE_28 = $14
+FILE_6C = $6C
+FILE_2C = $54
+FILE_64 = $74
+EASTER_EGG_FILENAME_1 equs "ぜるだ" ; Zelda
+EASTER_EGG_SONG_1 equ $60 ; MUSIC_ZELDA_NICKNAME_EASTER_EGG
+EASTER_EGG_FILENAME_2 equs "とたけけ" ; Totakeke
+EASTER_EGG_SONG_2 equ $3C ; MUSIC_TOTAKEKE_NICKNAME_EASTER_EGG
+THIEF_NAME equs "どろぼ－"
+DEBUG_SAVE_FILE_NAME equs "えすばはら"
```
Comment (kemenaran) [inline src/options.asm:91]: It's really nice to see the versions differences in this file :) 👍

File: src/code/map_loading.asm [hunk 5]
```diff
@@ -54,10 +54,16 @@ GetBGAttributesAddressForObject::
     and  a                                        ; $659A: $A7
     jp   z, .indoorPaletteEnd                     ; $659B: $CA $30 $66
 
+
     ld   a, BANK(IndoorsBGAttributesA)            ; $659E: $3E $23
     ldh  [hScratch8], a                           ; $65A0: $E0 $DF
+
+    ; hl += $200
+    ; Assertions were introduced in rgbds 0.4.0
+    ; ASSERT OverworldBGAttributesPointers + $200 == IndoorsBGAttributesPointers
```
Comment (kemenaran) [inline src/code/map_loading.asm:63]: I guess you can uncomment this line, now that the CI uses rgbds v0.4.0 :)

File: src/data/bg_attributes/overworld_banks.asm [hunk 6]
```diff
@@ -1,19 +1,17 @@
...
+    db BANK(Data_022_4000), BANK(Data_022_4000), BANK(Data_022_4C00), BANK(Data_022_4C00), BANK(Data_022_5400), BANK(Data_022_5400), BANK(Data_022_5400), BANK(Data_022_5400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6800), BANK(Data_022_7000), BANK(Data_022_6C00), BANK(Data_022_6C00)
+    db BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6800), BANK(Data_022_6800), BANK(Data_022_6C00), BANK(Data_022_6C00)
+    db BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_5000), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6400), BANK(Data_022_6800), BANK(Data_022_6800), BANK(Data_022_6C00), BANK(Data_022_6C00)
```
Comment (kemenaran) [inline src/data/bg_attributes/overworld_banks.asm:17]: ❤️

File: src/data/bg_attributes/overworld_pointers.asm [hunk 7]
```diff
@@ -1,265 +1,24 @@
...
+    dw Data_022_4000, Data_022_4000, Data_022_4C00, Data_022_4C00, Data_022_5400, Data_022_5400, Data_022_5400, Data_022_5400, Data_022_6000, Data_022_6000, Data_022_6400, Data_022_6400, Data_022_6800, Data_022_5C00, Data_022_6C00, Data_022_6C00
+    dw Data_022_4000, Data_022_4000, Data_022_4C00, Data_022_4C00, Data_022_5400, Data_022_5400, Data_022_5400, Data_022_5400, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6800, Data_022_7000, Data_022_6C00, Data_022_6C00
+    dw Data_022_5000, Data_022_5000, Data_022_5000, Data_022_5000, Data_022_5000, Data_022_5000, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6800, Data_022_6800, Data_022_6C00, Data_022_6C00
+    dw Data_022_5000, Data_022_5000, Data_022_5000, Data_022_5000, Data_022_5000, Data_022_5000, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6400, Data_022_6800, Data_022_6800, Data_022_6C00, Data_022_6C00
```
Comment (kemenaran) [inline src/data/bg_attributes/overworld_pointers.asm:19]: 💜

**Comment by kemenaran** [comment]
woohoo

**Pull request merged**